### ISSUE TITLE
Support `calc()` expressions in all properties

### ISSUE THREAD

**Issue by Jym77** [Siteimprove/alfa#1202]
New plan of attack… (post #1416 )

- [x] Replace the hybrid `Length` value in `alfa-style` with the one from `alfa-css/src/value/numeric`.
- [x] Switch all `Length` usage from `alfa-css/src/calculation/numeric` to `alfa-css/src/value/numeric`. This will probably require to provide a `parseBase` function in the later that can only parse non-calculated lengths (as a temporary solution). This will also need to expose all methods from the former on the later.
- [x] Wherever `Length` is used as a "final" value type (i.e. in `alfa-style`), replace `parseBase` with `parse` (i.e. accept calculation and provide a `resolve` helper). There may also be places in `alfa-media` that will need update. 
    Note: some style properties do not accept negative values, and this is treated differently for calculated and non-calculated values: [Range checking](https://drafts.csswg.org/css-values-4/#calc-range). 
    Note: this will likely require to buff up the new `Length` with methods from the old one (`isZero`, …)
    Note: it seems that calculation in media queries need to be parenthesised (`min-width: (10px + 10px)`, not `min-width: 10px + 10px`), need to find normative reference for that.
- [x] At this point, it should be possible to strip down `calculation/numeric/length` to only what is needed in math expressions.
- [x] Repeat the switch for other numeric types (number, integer, percentage, angle). 
    Note: it might be worth re-introducing a common `Dimension` abstraction on the new side also.
- [x] Add the `dimension-percentage` types we need, i.e. `length-percentage` and `angle-percentage`. 
    Note: it might be worth having an abstraction here also. 
    Note: it might be worth recording in the type whether the non-calculated value is holding a dimension or a percentage.
- [x] At this point, it should be possible to completely get rid of `alfa-style/src/property/value/compound.ts`.
- [ ] #1444 
- [x] Add type `Value.Parser<V> = Parser<Slice<Token>, V, string>` since we use it everywhere.
- [x] Clean up existing values to explicitly add `implements Resolvable` where relevant.
- [x] Update the other value types in `alfa-css` in order to use `src/value/numeric` rather than `src/calculation/numeric`. It is likely best to first introduce a `parseBase` to parse non-calculated values for easy migration.
  - [x] box
  - [x] collection
  - [x] color
  - [x] image
  - [x] keyword
  - [x] position
  - [x] shadow
  - [x] shape
  - [x] string
  - [x] transform
- [x] Strip down `src/calculation/numeric` to its bare minimum; remove `Math` inheritance from `Value` (rename `resolve2` to `resolve`); remove `src/calculation/numeric` inheritance from `Value`; move `src/value.ts` back in `src/value/value.ts`.
- [x] Clean up: use `Value.HasCalculation` directly in extends clause, instead of having an explicit `CALC` parameter.
- [x] Clean up the `parseBase` that shouldn't be needed anymore. 
    Note: need to check if all contexts actually allow calculations (e.g. media queries).
- [x] Export from `alfa-css` the `numeric` values rather than calculation; clean up any remaining "deep imports" to `@siteimprove/alfa-css/src/value/numeric`
- [x] ~Clean up the workarounds in R62 serialisation.~ #1426
- [x] Consider having a `Canonical` abstraction for all types, corresponding to the resolved and computed version of the type; maybe also hide away the `Calculated` an `Fixed` abstractions.
- [x] Lift `partiallyResolve` as a member method, returning `resolve()` by default, add it to `Resolvable` interface.
- [ ] Properly use `Resolvable` interface everywhere.
- [x] Clean up: use the parser helpers in tests.
- [ ] Clean up: make the sub-values properly `Serialisable` so that the correct JSON is used instead of top-level `Value.JSON`?
- [ ] Properly use the calculated value in "final" places (mostly `alfa-style`).
- [ ] 🎉 

---
After #980, easy support of `calc()` was added in #1201.
We now need to go through all properties and add support where relevant. Can also wait until the need arises…

- [x] background
- [x] background-attachment
- [x] background-clip
- [ ] background-color
- [x] background-image
- [x] background-origin
- [x] background-position
- [ ] background-position-x
- [ ] background-position-y
- [x] background-repeat
- [x] background-repeat-x
- [x] background-repeat-y
- [x] background-size
- [x] border
- [x] border-block
- [x] border-block-color
- [x] border-block-end
- [ ] border-block-end-color
- [x] border-block-end-style
- [x] border-block-end-width
- [x] border-block-start
- [ ] border-block-start-color
- [x] border-block-start-style
- [x] border-block-start-width
- [x] border-block-style
- [x] border-block-width
- [x] border-bottom
- [ ] border-bottom-color
- [x] border-bottom-left-radius
- [x] border-bottom-right-radius
- [x] border-bottom-style
- [x] border-bottom-width
- [x] border-collapse
- [x] border-color
- [x] border-end-end-radius
- [x] border-end-start-radius
- [x] border-image
- [ ] border-image-outset.ts
- [x] border-image-repeat
- [ ] border-image-slice
- [x] border-image-source
- [ ] border-image-width
- [x] border-inline
- [x] border-inline-color
- [x] border-inline-end
- [ ] border-inline-end-color
- [x] border-inline-end-style
- [x] border-inline-end-width
- [x] border-inline-start
- [ ] border-inline-start-color
- [x] border-inline-start-style
- [x] border-inline-start-width
- [x] border-inline-style
- [x] border-inline-width
- [x] border-left
- [ ] border-left-color
- [x] border-left-style
- [x] border-left-width
- [x] border-radius
- [x] border-right
- [ ] border-right-color
- [x] border-right-style
- [x] border-right-width
- [x] border-start-end-radius
- [x] border-start-start-radius
- [x] border-style
- [x] border-top
- [ ] border-top-color
- [x] border-top-left-radius
- [x] border-top-right-radius
- [x] border-top-style
- [x] border-top-width
- [x] border-width
- [x] bottom
- [ ] box-shadow
- [ ] clip
- [ ] clip-path
- [ ] color
- [ ] cursor
- [x] display
- [x] flex-direction
- [x] flex-flow
- [x] flex-wrap
- [x] float
- [x] font
- [x] font-family
- [X] font-size
- [ ] font-stretch
- [x] font-style
- [x] font-variant
- [x] font-variant-caps
- [x] font-variant-east-asian
- [x] font-variant-ligatures
- [x] font-variant-numeric
- [x] font-variant-position
- [x] font-weight
- [ ] height
- [x] inset
- [x] inset-block
- [x] inset-block-end
- [x] inset-block-start
- [x] inset-inline
- [x] inset-inline-end
- [x] inset-inline-start
- [x] left
- [x] letter-spacing
- [X] line-height
- [x] margin
- [x] margin-bottom
- [x] margin-left
- [x] margin-right
- [ ] margin-top
- [ ] min-height
- [ ] min-width
- [x] opacity
- [x] outline
- [ ] outline-color
- [x] outline-offset
- [x] outline-style
- [x] outline-width
- [x] overflow
- [x] overflow-x
- [x] overflow-y
- [x] position
- [x] right
- [ ] rotate
- [x] text-align
- [x] text-decoration
- [ ] text-decoration-color
- [x] text-decoration-line
- [x] text-decoration-style
- [x] text-decoration-thickness
- [x] text-indent
- [x] text-overflow
- [ ] text-shadow
- [x] text-transform
- [x] top
- [ ] transform
- [x] vertical-align
- [x] visibility
- [x] white-space
- [ ] width
- [x] word-spacing

---
- [x] Follow up on DEV-13509 (`outline-offset`, `outline-width`).
